### PR TITLE
[maxtext] Add block-causal attention

### PR DESCRIPTION
**Pull request by ethannnnnn**
## Motivation

Block diffusion needs attention that is causal across blocks while remaining
bidirectional within the block currently being corrupted or denoised. This is
an attention-kernel concern and can be reviewed independently of corruption,
training objectives, rollout, and RL.

## Scope

This PR adds only:

- an opt-in `block_diffusion` attention type and positive
  `causal_block_size` configuration;
- equivalent dense dot-product and TPU Tokamax Splash block-causal masks;
- load-balanced context-parallel mask support;
- validation that rejects packing and unsupported attention/hardware
  combinations; and
- focused mask, configuration, and autoregressive-regression tests.

It does not depend on the block-diffusion primitive PR, corrupt training data,
change a loss function, add a model-specific branch, or enable diffusion by
default.

## Mask semantics

For query position `q` and key position `k`, attention is allowed when

```text
floor(q / block_size) >= floor(k / block_size)
```

Therefore, all positions in the current block can attend to one another and to
all earlier blocks, but no earlier block can attend to a later block. Segment
boundaries remain enforced. Packing is rejected because document-local block
origins are not yet represented by the attention metadata. Autoregressive
decode keeps the existing causal mask path unchanged.

## Compatibility

The default remains `attention_type=global`. The new behavior is activated only
when `attention_type=block_diffusion` is selected explicitly. Specialized
per-layer attention declarations continue to override the global configuration.

This implementation remains code-independent from the primitives landed in
#4737: it does not import or activate them. The branch is rebased onto the
current `main` containing #4737 so reviewers see only this attention change.

## Tests

```text
python -m pytest -q \
  tests/unit/attention_test.py \
  tests/unit/configs_value_test.py \
  tests/unit/tokamax_splash_attention_mask_test.py

76 passed, 48 skipped, 36 subtests passed
```

The skipped cases require unavailable platform-specific kernels. Pyink, Python
compilation, and `git diff --check` also pass for the changed files.

Focused branch coverage for the changed executable code in
`src/maxtext/layers/attention_op.py` is complete: 53/53 changed lines and
26/26 changed branches are covered.

## Checklist

- [x] I have performed a self-review of my code.
- [x] I added comments or docstrings for the non-obvious block-mask invariants.
- [x] I added focused dense, Splash, load-balanced, configuration, and autoregressive-regression tests.
- [x] End-to-end model training is not applicable to this default-off attention layer; the affected CPU suites are listed above.
- [x] No standalone documentation page or toctree change is needed; the public design document is linked below.

Design document:
https://docs.google.com/document/d/1N7KcCoAIErB2CV9EJ2G1u_mdN-AQgqwNUYSvM0mtqMI/edit

### CODE REVIEW

File: src/maxtext/layers/attention_op.py [hunk 1]
```diff
@@ -234,6 +288,18 @@ def _generate_chunk_attention_mask(mask_shape: tuple[int, int], chunk_size: int,
   return chunk_mask
 
 
+def _generate_block_causal_attention_mask(
+    mask_shape: tuple[int, int], causal_block_size: int, q_offset: int = 0
+) -> jax.Array:
+  """Generates a block-causal mask, bidirectional within each block."""
```
Comment (shralex) [inline src/maxtext/layers/attention_op.py]: Suggestion: 
  
"""Generates a block-causal mask for Block Diffusion Language Models (BD3LMs).
  Implements the block-causal attention pattern (M_BC) described in
  Arriola et al., "Block Diffusion: Interpolating Between Autoregressive and
  Diffusion Language Models" (https://arxiv.org/abs/2503.09573).
  For block size B = causal_block_size, query index q, and key index k:
    - Within block i (floor(q/B) == floor(k/B)): tokens attend bidirectionally.
    - Across blocks (floor(q/B) > floor(k/B)): block i attends causally to
      all preceding blocks 0, 1, ..., i-1, but not to future blocks.
  Note:
    - B = 1 reduces to standard autoregressive causal attention.
    - B = L (sequence length) reduces to full bidirectional attention.
  """

File: src/maxtext/configs/types.py [hunk 2]
```diff
@@ -3484,6 +3488,16 @@ def calculate_global_batch_sizes(per_device_batch_size, expansion_factor, num_de
         not isinstance(self.sliding_window_size, int) or self.sliding_window_size <= 0
     ):
       raise ValueError("`sliding_window_size` must be an integer > 0 for 'local_sliding' attention.")
+    if self.attention_type == AttentionType.BLOCK_DIFFUSION.value:
+      if self.packing:
```
Comment (shralex) [inline src/maxtext/configs/types.py:3508]: lets explain a bit better why its an error, like:

        # Document-local block origins inside a packed sequence are not yet
        # tracked in attention metadata; packing without realignment would cause
        # cross-document block-attention leakage.

File: src/maxtext/layers/attentions.py [hunk 3]
```diff
@@ -120,7 +120,7 @@ def attention_as_linen(
     float32_logits: bool = False,  # cast logits in float32 for stability.
     quant: Optional[Quant] = None,
     kv_quant: Optional[KVQuant] = None,
-    attention_type: AttentionType = AttentionType.GLOBAL,  # Default to global attention
+    attention_type: AttentionType | None = None,
```
Comment (entrpn) [inline src/maxtext/layers/attentions.py]: why is it necessary to change this default value from GLOBAL to none?

Comment (ethannnnnn) [reply src/maxtext/layers/attentions.py]: Good catch. None became unnecessary after adding _resolve_attention_type. I restored the existing AttentionType.GLOBAL default in both constructors. Configured block diffusion still activates through the resolver.

File: src/maxtext/layers/attention_op.py [hunk 4]
```diff
@@ -862,6 +935,19 @@ def generate_attention_mask(
         )
       output_mask = chunk_mask * output_mask
 
+    elif self.attention_type == AttentionType.BLOCK_DIFFUSION and model_mode != MODEL_MODE_AUTOREGRESSIVE:
```
Comment (shralex) [inline src/maxtext/layers/attention_op.py:1000]: Lets clarify why, e.g.:

    # For standard token-by-token autoregressive decoding (MODEL_MODE_AUTOREGRESSIVE),
    # keep the existing causal mask path unchanged. BD3LM generation operates via
    # block-level parallel sampling rather than token-level autoregressive decode.

File: src/maxtext/configs/base.yml [hunk 5]
```diff
@@ -387,12 +387,13 @@ param_scan_axis: 1
...
 # The attention_type parameter determines the variants of attention, e.g. global or local_sliding
 attention: 'autoselected' # Supported attention: autoselected, dot_product, flash, cudnn_flash_te
-attention_type: 'global' # Supported attention_type: global, local_sliding, chunk, mla
+attention_type: 'global' # Supported attention_type: global, local_sliding, chunk, mla, full, compressed, block_diffusion
 share_kv_projections: false # Note: Not compatible with attention_type='mla'
 attention_bias: false # If true, adds a learnable bias to the query, key, and value projections
 attention_sink: false
 sliding_window_size: 0
 chunk_attn_window_size: 0
+causal_block_size: 32 # Number of token positions in each bidirectional block for block-causal attention
```
Comment (shralex) [inline src/maxtext/configs/base.yml]: Suggest the following comment instead, so users can find the paper without digging into attntions.py

# Token block size B for block-causal attention in Block Diffusion (arXiv:2503.09573). Tokens attend bidirectionally within a block and causally across blocks.

**Review by shralex — APPROVED** [review]
Thank  you for the contribution! Added a few documentation suggestions to make it easier to navigate for future users / maintainers.

File: src/maxtext/layers/attention_op.py (hunk 1 above)
Comment (ethannnnnn) [reply src/maxtext/layers/attention_op.py]: Done. The mask helper docstring now cites the Block Diffusion paper, defines within-block and across-block attention, and documents the `B = 1` causal and `B = L` bidirectional limits.

File: src/maxtext/configs/types.py (hunk 2 above)
Comment (ethannnnnn) [reply src/maxtext/configs/types.py:3508]: Done. The validation comment now explains that document-local block origins are not tracked in attention metadata and that packing without realignment could leak block attention across documents.

File: src/maxtext/layers/attention_op.py (hunk 4 above)
Comment (ethannnnnn) [reply src/maxtext/layers/attention_op.py:1000]: Done. The comment now clarifies that standard autoregressive decoding preserves the causal mask path, while BD3LM generation uses block-level parallel sampling.

File: src/maxtext/configs/base.yml (hunk 5 above)
Comment (ethannnnnn) [reply src/maxtext/configs/base.yml]: Done. The `causal_block_size` configuration comment now links the Block Diffusion paper and summarizes within-block bidirectional and across-block causal attention.

**Review by entrpn — APPROVED** [review]

**Review by hengtaoguo — APPROVED** [review]
The github actions are temporarily down so the PR tests are queued. Let's watch out for this.

File: tests/unit/configs_value_test.py [hunk 6]
```diff
@@ -299,6 +300,57 @@ def test_load_balanced_chunk_context_parallel_config(self):
...
+            _BASE_CONFIG_PATH,
+            "run_name=test",
+            "steps=1",
+            "attention=dot_product",
+            "attention_type=block_diffusion",
+            "causal_block_size=7",
+            "vocab_size=256",
+            "max_target_length=2048",
+            "hardware=cpu",
+            "packing=False",
```
Comment (hengtaoguo) [inline tests/unit/configs_value_test.py:316]: I'm seeing an internal test failure related to jax distributed initialization. So allow me add a new flag here to assist the bug fix `skip_jax_distributed_system=True`.

cc @entrpn The error message is [here](https://fusion2.corp.google.com/presubmit/960733080/OCL:960733080:BASE:960733443:1786085015733:60c5d675/targets/invocations/1bff1b30-8f67-413b-9001-16bca4bc6793/targets/%2F%2Fthird_party%2Fpy%2Fmaxtext:tests%2Funit%2Fconfigs_value_test/tests).

File: tests/unit/configs_value_test.py [hunk 7]
```diff
@@ -299,6 +300,57 @@ def test_load_balanced_chunk_context_parallel_config(self):
...
+    base_overrides = {
+        "run_name": "test",
+        "steps": 1,
+        "attention": "dot_product",
+        "attention_type": "block_diffusion",
+        "causal_block_size": 32,
+        "vocab_size": 256,
+        "max_target_length": 2048,
+        "hardware": "cpu",
+        "packing": False,
```
Comment (hengtaoguo) [inline tests/unit/configs_value_test.py:335]: Add `"skip_jax_distributed_system": True` here too.

**Review by hengtaoguo — APPROVED** [review]
